smart-hadoop-support/smart-hadoop/src/main/java/org/smartdata/hdfs/scheduler/SmallFileScheduler.java

```diff
@@ -0,0 +1,512 @@
+    if (COMPACT_ACTION_NAME.equals(actionInfo.getActionName())) {
+      // Get container file and small file list of this action
+      String containerFilePath = actionInfo.getArgs().get(
+          SmallFileCompactAction.CONTAINER_FILE);
+      String smallFiles = actionInfo.getArgs().get(HdfsAction.FILE_PATH);
+      ArrayList<String> smallFileList = new Gson().fromJson(
+          smallFiles, new TypeToken<ArrayList<String>>() {
+          }.getType());
+
+      // Check if container file is locked and retry
+      if (containerFileLock.containsKey(containerFilePath)) {
+        return ScheduleResult.RETRY;
+      } else {
+        // Check if small file path is valid and unlocked
+        for (String smallFile : smallFileList) {
+          if (smallFile == null || smallFile.isEmpty()) {
+            LOG.debug("Illegal small file path: " + smallFile);
+            actionInfo.setResult("Illegal small file path: " + smallFile);
+            return ScheduleResult.FAIL;
+          } else if (compactSmallFileLock.contains(smallFile)) {
+            String errMsg = String.format("%s is locked.", smallFile);
+            LOG.debug(errMsg);
+            actionInfo.setResult(errMsg);
+            return ScheduleResult.FAIL;
+          } else if (containerFileCache.contains(smallFile)) {
+            String errMsg = String.format("%s is not small file.", smallFile);
+            LOG.debug(errMsg);
+            actionInfo.setResult(errMsg);
+            return ScheduleResult.FAIL;
+          }
+        }
+
+        // Get small file state map from meta store.
+        Map<String, FileState> fileStateMap;
+        try {
+          fileStateMap = metaStore.getFileStates(smallFileList);
+        } catch (MetaStoreException e) {
+          LOG.error("Failed to get file states of small files. " + e.toString());
+          actionInfo.setResult(
+              "Failed to get file states of small files. " + e.toString());
+          return ScheduleResult.FAIL;
+        }
+
+        // Check if the state of small file is NORMAL
+        if (fileStateMap.size() != 0) {
+          for (String smallFile : smallFileList) {
+            if (fileStateMap.containsKey(smallFile)) {
+              FileState.FileType smallFileType = fileStateMap.get(
+                  smallFile).getFileType();
+              if (smallFileType != FileState.FileType.NORMAL) {
+                String errMsg = String.format(
+                    "%s has invalid file state %s for small file compact.",
+                    smallFile, smallFileType.toString());
+                LOG.debug(errMsg);
+                actionInfo.setResult(errMsg);
+                return ScheduleResult.FAIL;
+              }
+            }
+          }
+        }
+
+        try {
+          // Lock container file and set whether exists already
+          // Lock small files
+          boolean exist = dfsClient.exists(containerFilePath);
```

Comment:
This part of work can be shifted to action, then scheduler does not have to delete the container file either.